routing/index_graph_starter.cpp

```diff
@@ -1,180 +1,396 @@
+    FakeVertex forwardPartOfReal(isStart ? projection.m_junction : backJunction,
+                                 isStart ? frontJunction : projection.m_junction,
+                                 FakeVertex::Type::PartOfReal);
+    auto const fakeForwardSegment = m_fake.GetSegment(forwardPartOfReal, fakeNumerationStart);
+    m_fake.AddFakeVertex(projectionSegment, fakeForwardSegment, forwardPartOfReal,
+                         isStart /* isOutgoing */, true /* isPartOfReal */, projection.m_segment);
 
-IndexGraphStarter::IndexGraphStarter(FakeVertex const & start, FakeVertex const & finish,
-                                     WorldGraph & graph)
+    bool const oneWay =
+        m_graph
+            .GetRoadGeometry(projection.m_segment.GetMwmId(), projection.m_segment.GetFeatureId())
+            .IsOneWay();
+    if (!strictForward && !oneWay)
+    {
+      auto const backwardSegment = GetReverseSegment(projection.m_segment);
+      FakeVertex backwardPartOfReal(isStart ? projection.m_junction : frontJunction,
+                                    isStart ? backJunction : projection.m_junction,
+                                    FakeVertex::Type::PartOfReal);
+      auto const fakeBackwardSegment = m_fake.GetSegment(backwardPartOfReal, fakeNumerationStart);
+      m_fake.AddFakeVertex(projectionSegment, fakeBackwardSegment, backwardPartOfReal,
+                           isStart /* isOutgoing */, true /* isPartOfReal */, backwardSegment);
+    }
+  }
+}
+
+void IndexGraphStarter::AddStart(FakeEnding const & startEnding, FakeEnding const & finishEnding,
+                                 bool strictForward, uint32_t & fakeNumerationStart)
+{
+  AddEnding(startEnding, finishEnding, true /* isStart */, strictForward, fakeNumerationStart);
+}
+
+void IndexGraphStarter::AddFinish(FakeEnding const & finishEnding, FakeEnding const & startEnding,
+                                  uint32_t & fakeNumerationStart)
+{
+  AddEnding(finishEnding, startEnding, false /* isStart */, false /* strictForward */,
+            fakeNumerationStart);
+}
+
+IndexGraphStarter::IndexGraphStarter(FakeEnding const & startEnding,
+                                     FakeEnding const & finishEnding, uint32_t fakeNumerationStart,
+                                     bool strictForward, WorldGraph & graph)
   : m_graph(graph)
-  , m_start(start.GetSegment(),
-            CalcProjectionToSegment(start.GetSegment(), start.GetJunction(), graph),
-            start.GetStrictForward())
-  , m_finish(finish.GetSegment(),
-             CalcProjectionToSegment(finish.GetSegment(), finish.GetJunction(), graph),
-             finish.GetStrictForward())
 {
+  m_startId = fakeNumerationStart;
+  AddStart(startEnding, finishEnding, strictForward, fakeNumerationStart);
+  m_finishId = fakeNumerationStart;
+  AddFinish(finishEnding, startEnding, fakeNumerationStart);
+}
+
+void IndexGraphStarter::Append(FakeEdgesContainer const & container)
+{
+  m_finishId = container.m_finishId;
```

Comment:
Why not to:
```c++
m_finishId = max(m_finishId, container.m_finishId);
```
?

Reply:
сейчас к id нет никаких требований кроме уникальности (нумеровать можно сверху вниз, снизу вверх ещё как-то), не вижу особого смысла их добавлять

Reply:
But what if before the assignment container.m_finishId is less than this->m_finishId?

Reply:
The only requirement is all fake vertex segmentIds must be different. We never compare fake segmentIds. I can check all Ids are different but it will cost n*logn.

Reply:
I added m_startId to FakeEdges container and CHECK_LESS(m_finishId, container.m_startId).

Reply:
Поняла что это неправильный чек, например в AdjustRoute у нас starter для начального куска маршрута конструируется позже чем FakeEdgesContainer и соотвтетствено имеет номера вершин больше чем в FakeEdgesContainer. Уберу CHECK_LESS, добавлю чек на уникальность вершин, это является реальным требованием.

Reply:
сделала чек на уникальность fake segment id